### ISSUE TITLE
Something wrong with desired MSE after move to doublefann

### ISSUE THREAD

**Issue by bararchy** [bararchy/crystal-fann#6]
After moving to use `doublefann` so that input parameters can be `Float64` we got some bug with the MSE values, running specs shows

```
........Max epochs     8000. Desired error: -0.0000000000.
Epochs            1. Current error: 0.2504344881. Bit fail 4.
Epochs         1000. Current error: 0.0000258677. Bit fail 0.
Epochs         2000. Current error: 0.0000260366. Bit fail 0.
Epochs         3000. Current error: 0.0000253705. Bit fail 0.
Epochs         4000. Current error: 0.0000244194. Bit fail 0.
Epochs         5000. Current error: 0.0000170922. Bit fail 0.
Epochs         6000. Current error: 0.0000164790. Bit fail 0.
Epochs         7000. Current error: 0.0000157123. Bit fail 0.
Epochs         8000. Current error: 0.0000215559. Bit fail 0.
.Max neurons 500. Desired error: -0.000000
```

Where relevant parts are:

```
Desired error: -0.0000000000
Desired error: -0.000000
```

Those should have used the configured amount of:
```
0.001
and
0.1
```

**Comment by bararchy** [comment]
Still digging, it seems that sometimes the MSE would also be `Desired error: -518969491456.0000000000.`

**Comment by bararchy** [comment]
Further digging into this, makes me think  that it might be a local issue, pausing work until someone can test master too.

Update -- Seems to fail for me on 8b95a2b2ec9018058a0a0cc0ffc7b1d4ca1dbebb too, so not something related to latest changes

**Comment by hugoabonizio** [comment]
Specs seems to be randomly failing, here are two consecutive runs:

```
$ crystal spec
......Max epochs     8000. Desired error: -518969491456.0000000000.
Epochs            1. Current error: 0.2500234544. Bit fail 4.
Epochs         1000. Current error: 0.1806672662. Bit fail 3.
Epochs         2000. Current error: 0.1792608500. Bit fail 3.
Epochs         3000. Current error: 0.1790057719. Bit fail 3.
Epochs         4000. Current error: 0.1789760590. Bit fail 3.
Epochs         5000. Current error: 0.1789408326. Bit fail 3.
Epochs         6000. Current error: 0.1789273918. Bit fail 3.
Epochs         7000. Current error: 0.1789356917. Bit fail 3.
Epochs         8000. Current error: 0.1789264679. Bit fail 3.
FMax neurons 500. Desired error: -0.000000
Neurons       0. Current error: 0.250001. Total error:  1.0000. Epochs    51. Bit fail   4
Neurons       9. Current error: 0.000000. Total error:  0.0000. Epochs  1777. Bit fail   0. candidate steepness 0.75. function FANN_COS_SYMMETRIC
Neurons      19. Current error: 0.000000. Total error:  0.0000. Epochs  3848. Bit fail   0. candidate steepness 0.75. function FANN_SIN
Neurons      29. Current error: 0.000000. Total error:  0.0000. Epochs  5870. Bit fail   0. candidate steepness 0.25. function FANN_COS
Neurons      39. Current error: 0.000000. Total error:  0.0000. Epochs  7893. Bit fail   0. candidate steepness 0.75. function FANN_COS_SYMMETRIC
Neurons      49. Current error: 0.000000. Total error:  0.0000. Epochs  9925. Bit fail   0. candidate steepness 0.25. function FANN_COS_SYMMETRIC
Neurons      59. Current error: 0.000000. Total error:  0.0000. Epochs 11949. Bit fail   0. candidate steepness 1.00. function FANN_SIN_SYMMETRIC
Neurons      69. Current error: 0.000000. Total error:  0.0000. Epochs 13975. Bit fail   0. candidate steepness 1.00. function FANN_SIGMOID
Neurons      79. Current error: 0.000000. Total error:  0.0000. Epochs 15997. Bit fail   0. candidate steepness 0.50. function FANN_SIGMOID
Neurons      89. Current error: 0.000000. Total error:  0.0000. Epochs 18007. Bit fail   0. candidate steepness 0.50. function FANN_GAUSSIAN
Neurons      99. Current error: 0.000000. Total error:  0.0000. Epochs 20037. Bit fail   0. candidate steepness 0.50. function FANN_SIN_SYMMETRIC
Neurons     109. Current error: 0.000000. Total error:  0.0000. Epochs 22063. Bit fail   0. candidate steepness 0.50. function FANN_SIN
Neurons     119. Current error: 0.000000. Total error:  0.0000. Epochs 24082. Bit fail   0. candidate steepness 1.00. function FANN_COS_SYMMETRIC
Neurons     129. Current error: 0.000000. Total error:  0.0000. Epochs 26106. Bit fail   0. candidate steepness 0.25. function FANN_SIGMOID
Neurons     139. Current error: 0.000000. Total error:  0.0000. Epochs 28122. Bit fail   0. candidate steepness 0.75. function FANN_SIN
Neurons     149. Current error: 0.000000. Total error:  0.0000. Epochs 30139. Bit fail   0. candidate steepness 0.75. function FANN_SIGMOID
Neurons     159. Current error: 0.000000. Total error:  0.0000. Epochs 32161. Bit fail   0. candidate steepness 0.75. function FANN_COS_SYMMETRIC
Neurons     169. Current error: 0.000000. Total error:  0.0000. Epochs 34220. Bit fail   0. candidate steepness 0.50. function FANN_SIGMOID
Neurons     179. Current error: 0.000000. Total error:  0.0000. Epochs 36259. Bit fail   0. candidate steepness 0.50. function FANN_COS
Neurons     189. Current error: 0.000000. Total error:  0.0000. Epochs 38275. Bit fail   0. candidate steepness 0.75. function FANN_COS
Neurons     199. Current error: 0.000000. Total error:  0.0000. Epochs 40301. Bit fail   0. candidate steepness 1.00. function FANN_SIN
Neurons     209. Current error: 0.000000. Total error:  0.0000. Epochs 42314. Bit fail   0. candidate steepness 1.00. function FANN_COS
Neurons     219. Current error: 0.000000. Total error:  0.0000. Epochs 44333. Bit fail   0. candidate steepness 0.25. function FANN_COS_SYMMETRIC
Neurons     229. Current error: 0.000000. Total error:  0.0000. Epochs 46345. Bit fail   0. candidate steepness 0.75. function FANN_SIGMOID
Neurons     239. Current error: 0.000000. Total error:  0.0000. Epochs 48358. Bit fail   0. candidate steepness 0.50. function FANN_SIN
Neurons     249. Current error: 0.000000. Total error:  0.0000. Epochs 50368. Bit fail   0. candidate steepness 0.50. function FANN_SIN_SYMMETRIC
Neurons     259. Current error: 0.000000. Total error:  0.0000. Epochs 52389. Bit fail   0. candidate steepness 0.25. function FANN_SIN_SYMMETRIC
Neurons     269. Current error: 0.000000. Total error:  0.0000. Epochs 54407. Bit fail   0. candidate steepness 0.75. function FANN_SIN
Neurons     279. Current error: 0.000000. Total error:  0.0000. Epochs 56417. Bit fail   0. candidate steepness 0.50. function FANN_COS_SYMMETRIC
Neurons     289. Current error: 0.000000. Total error:  0.0000. Epochs 58430. Bit fail   0. candidate steepness 1.00. function FANN_SIN_SYMMETRIC
Neurons     299. Current error: 0.000000. Total error:  0.0000. Epochs 60446. Bit fail   0. candidate steepness 0.75. function FANN_SIGMOID
Neurons     309. Current error: 0.000000. Total error:  0.0000. Epochs 62456. Bit fail   0. candidate steepness 0.75. function FANN_COS_SYMMETRIC
Neurons     319. Current error: 0.000000. Total error:  0.0000. Epochs 64466. Bit fail   0. candidate steepness 0.75. function FANN_COS_SYMMETRIC
Neurons     329. Current error: 0.000000. Total error:  0.0000. Epochs 66488. Bit fail   0. candidate steepness 1.00. function FANN_COS
Neurons     339. Current error: 0.000000. Total error:  0.0000. Epochs 68511. Bit fail   0. candidate steepness 0.50. function FANN_SIN
Neurons     349. Current error: 0.000000. Total error:  0.0000. Epochs 70521. Bit fail   0. candidate steepness 1.00. function FANN_SIGMOID
Neurons     359. Current error: 0.000000. Total error:  0.0000. Epochs 72531. Bit fail   0. candidate steepness 0.50. function FANN_SIGMOID
Neurons     369. Current error: 0.000000. Total error:  0.0000. Epochs 74544. Bit fail   0. candidate steepness 0.50. function FANN_SIGMOID
Neurons     379. Current error: 0.000000. Total error:  0.0000. Epochs 76556. Bit fail   0. candidate steepness 0.25. function FANN_COS_SYMMETRIC
Neurons     389. Current error: 0.000000. Total error:  0.0000. Epochs 78570. Bit fail   0. candidate steepness 1.00. function FANN_COS_SYMMETRIC
Neurons     399. Current error: 0.000000. Total error:  0.0000. Epochs 80581. Bit fail   0. candidate steepness 0.75. function FANN_SIGMOID
Neurons     409. Current error: 0.000000. Total error:  0.0000. Epochs 82596. Bit fail   0. candidate steepness 0.50. function FANN_SIGMOID
Neurons     419. Current error: 0.000000. Total error:  0.0000. Epochs 84609. Bit fail   0. candidate steepness 1.00. function FANN_SIN_SYMMETRIC
Neurons     429. Current error: 0.000000. Total error:  0.0000. Epochs 86619. Bit fail   0. candidate steepness 0.25. function FANN_SIN_SYMMETRIC
Neurons     439. Current error: 0.000000. Total error:  0.0000. Epochs 88648. Bit fail   0. candidate steepness 0.50. function FANN_SIGMOID
Neurons     449. Current error: 0.000000. Total error:  0.0000. Epochs 90667. Bit fail   0. candidate steepness 0.75. function FANN_COS_SYMMETRIC
Neurons     459. Current error: 0.000000. Total error:  0.0000. Epochs 92680. Bit fail   0. candidate steepness 1.00. function FANN_COS
Neurons     469. Current error: 0.000000. Total error:  0.0000. Epochs 94694. Bit fail   0. candidate steepness 0.25. function FANN_SIGMOID
Neurons     479. Current error: 0.000000. Total error:  0.0000. Epochs 96711. Bit fail   0. candidate steepness 0.50. function FANN_SIGMOID
Neurons     489. Current error: 0.000000. Total error:  0.0000. Epochs 98725. Bit fail   0. candidate steepness 0.50. function FANN_COS_SYMMETRIC
Neurons     499. Current error: 0.000000. Total error:  0.0000. Epochs 100735. Bit fail   0. candidate steepness 0.50. function FANN_SIN_SYMMETRIC
Train outputs    Current error: 0.000000. Epochs 100936
...

Failures:

  1) Fann::Network train on batch
     Failure/Error: (result < [0.1]).should be_true

       Expected: true
            got: false

     # spec/network_spec.cr:58

Finished in 43.69 seconds
10 examples, 1 failures, 0 errors, 0 pending

Failed examples:

crystal spec spec/network_spec.cr:44 # Fann::Network train on batch
```

```
$ crystal spec
.....FMax epochs     8000. Desired error: -518969491456.0000000000.
Epochs            1. Current error: 0.2500280142. Bit fail 4.
Epochs         1000. Current error: 0.2500908077. Bit fail 1.
Epochs         2000. Current error: 0.2500316501. Bit fail 1.
Epochs         3000. Current error: 0.2501344681. Bit fail 1.
Epochs         4000. Current error: 0.2503082454. Bit fail 1.
Epochs         5000. Current error: 0.2505527437. Bit fail 1.
Epochs         6000. Current error: 0.2502878308. Bit fail 1.
Epochs         7000. Current error: 0.2500071526. Bit fail 1.
Epochs         8000. Current error: 0.2500744462. Bit fail 1.
.Max neurons 500. Desired error: -0.000000
Neurons       0. Current error: 0.250000. Total error:  1.0000. Epochs    51. Bit fail   4
Neurons       9. Current error: 0.000000. Total error:  0.0000. Epochs  1764. Bit fail   0. candidate steepness 0.50. function FANN_SIN_SYMMETRIC
Neurons      19. Current error: 0.000000. Total error:  0.0000. Epochs  3802. Bit fail   0. candidate steepness 0.25. function FANN_COS_SYMMETRIC
Neurons      29. Current error: 0.000000. Total error:  0.0000. Epochs  5835. Bit fail   0. candidate steepness 0.50. function FANN_COS_SYMMETRIC
Neurons      39. Current error: 0.000000. Total error:  0.0000. Epochs  7858. Bit fail   0. candidate steepness 0.50. function FANN_SIN_SYMMETRIC
Neurons      49. Current error: 0.000000. Total error:  0.0000. Epochs  9870. Bit fail   0. candidate steepness 1.00. function FANN_COS_SYMMETRIC
Neurons      59. Current error: 0.000000. Total error:  0.0000. Epochs 11882. Bit fail   0. candidate steepness 1.00. function FANN_SIGMOID
Neurons      69. Current error: 0.000000. Total error:  0.0000. Epochs 13902. Bit fail   0. candidate steepness 0.75. function FANN_SIN
Neurons      79. Current error: 0.000000. Total error:  0.0000. Epochs 15916. Bit fail   0. candidate steepness 0.50. function FANN_COS_SYMMETRIC
Neurons      89. Current error: 0.000000. Total error:  0.0000. Epochs 17934. Bit fail   0. candidate steepness 0.75. function FANN_SIGMOID
Neurons      99. Current error: 0.000000. Total error:  0.0000. Epochs 19954. Bit fail   0. candidate steepness 1.00. function FANN_SIN_SYMMETRIC
Neurons     109. Current error: 0.000000. Total error:  0.0000. Epochs 21971. Bit fail   0. candidate steepness 0.50. function FANN_SIN
Neurons     119. Current error: 0.000000. Total error:  0.0000. Epochs 24000. Bit fail   0. candidate steepness 0.50. function FANN_SIN_SYMMETRIC
Neurons     129. Current error: 0.000000. Total error:  0.0000. Epochs 26014. Bit fail   0. candidate steepness 0.75. function FANN_SIN_SYMMETRIC
Neurons     139. Current error: 0.000000. Total error:  0.0000. Epochs 28042. Bit fail   0. candidate steepness 0.50. function FANN_SIGMOID
Neurons     149. Current error: 0.000000. Total error:  0.0000. Epochs 30058. Bit fail   0. candidate steepness 0.25. function FANN_SIGMOID
Neurons     159. Current error: 0.000000. Total error:  0.0000. Epochs 32080. Bit fail   0. candidate steepness 1.00. function FANN_COS_SYMMETRIC
Neurons     169. Current error: 0.000000. Total error:  0.0000. Epochs 34094. Bit fail   0. candidate steepness 0.25. function FANN_SIN_SYMMETRIC
Neurons     179. Current error: 0.000000. Total error:  0.0000. Epochs 36119. Bit fail   0. candidate steepness 0.50. function FANN_SIGMOID
Neurons     189. Current error: 0.000000. Total error:  0.0000. Epochs 38147. Bit fail   0. candidate steepness 0.75. function FANN_SIGMOID
Neurons     199. Current error: 0.000000. Total error:  0.0000. Epochs 40159. Bit fail   0. candidate steepness 0.75. function FANN_COS_SYMMETRIC
Neurons     209. Current error: 0.000000. Total error:  0.0000. Epochs 42179. Bit fail   0. candidate steepness 0.75. function FANN_SIGMOID
Neurons     219. Current error: 0.000000. Total error:  0.0000. Epochs 44200. Bit fail   0. candidate steepness 0.25. function FANN_SIN_SYMMETRIC
Neurons     229. Current error: 0.000000. Total error:  0.0000. Epochs 46217. Bit fail   0. candidate steepness 1.00. function FANN_COS_SYMMETRIC
Neurons     239. Current error: 0.000000. Total error:  0.0000. Epochs 48231. Bit fail   0. candidate steepness 0.75. function FANN_SIGMOID
Neurons     249. Current error: 0.000000. Total error:  0.0000. Epochs 50275. Bit fail   0. candidate steepness 0.25. function FANN_SIGMOID
Neurons     259. Current error: 0.000000. Total error:  0.0000. Epochs 52285. Bit fail   0. candidate steepness 0.75. function FANN_SIGMOID
Neurons     269. Current error: 0.000000. Total error:  0.0000. Epochs 54310. Bit fail   0. candidate steepness 0.75. function FANN_SIN
Neurons     279. Current error: 0.000000. Total error:  0.0000. Epochs 56320. Bit fail   0. candidate steepness 0.75. function FANN_COS_SYMMETRIC
Neurons     289. Current error: 0.000000. Total error:  0.0000. Epochs 58330. Bit fail   0. candidate steepness 1.00. function FANN_COS_SYMMETRIC
Neurons     299. Current error: 0.000000. Total error:  0.0000. Epochs 60340. Bit fail   0. candidate steepness 0.50. function FANN_SIGMOID
Neurons     309. Current error: 0.000000. Total error:  0.0000. Epochs 62359. Bit fail   0. candidate steepness 0.75. function FANN_SIN
Neurons     319. Current error: 0.000000. Total error:  0.0000. Epochs 64373. Bit fail   0. candidate steepness 0.25. function FANN_ELLIOT_SYMMETRIC
Neurons     329. Current error: 0.000000. Total error:  0.0000. Epochs 66388. Bit fail   0. candidate steepness 0.50. function FANN_SIGMOID
Neurons     339. Current error: 0.000000. Total error:  0.0000. Epochs 68401. Bit fail   0. candidate steepness 1.00. function FANN_SIGMOID
Neurons     349. Current error: 0.000000. Total error:  0.0000. Epochs 70414. Bit fail   0. candidate steepness 0.25. function FANN_SIN_SYMMETRIC
Neurons     359. Current error: 0.000000. Total error:  0.0000. Epochs 72440. Bit fail   0. candidate steepness 0.50. function FANN_COS_SYMMETRIC
Neurons     369. Current error: 0.000000. Total error:  0.0000. Epochs 74458. Bit fail   0. candidate steepness 0.75. function FANN_COS_SYMMETRIC
Neurons     379. Current error: 0.000000. Total error:  0.0000. Epochs 76468. Bit fail   0. candidate steepness 0.75. function FANN_COS
Neurons     389. Current error: 0.000000. Total error:  0.0000. Epochs 78480. Bit fail   0. candidate steepness 0.75. function FANN_COS_SYMMETRIC
Neurons     399. Current error: 0.000000. Total error:  0.0000. Epochs 80497. Bit fail   0. candidate steepness 0.25. function FANN_COS
Neurons     409. Current error: 0.000000. Total error:  0.0000. Epochs 82509. Bit fail   0. candidate steepness 1.00. function FANN_SIGMOID
Neurons     419. Current error: 0.000000. Total error:  0.0000. Epochs 84534. Bit fail   0. candidate steepness 0.25. function FANN_SIN_SYMMETRIC
Neurons     429. Current error: 0.000000. Total error:  0.0000. Epochs 86549. Bit fail   0. candidate steepness 1.00. function FANN_SIN
Neurons     439. Current error: 0.000000. Total error:  0.0000. Epochs 88566. Bit fail   0. candidate steepness 0.75. function FANN_SIGMOID
Neurons     449. Current error: 0.000000. Total error:  0.0000. Epochs 90576. Bit fail   0. candidate steepness 0.25. function FANN_ELLIOT_SYMMETRIC
Neurons     459. Current error: 0.000000. Total error:  0.0000. Epochs 92593. Bit fail   0. candidate steepness 0.75. function FANN_SIGMOID
Neurons     469. Current error: 0.000000. Total error:  0.0000. Epochs 94603. Bit fail   0. candidate steepness 0.25. function FANN_SIN
Neurons     479. Current error: 0.000000. Total error:  0.0000. Epochs 96622. Bit fail   0. candidate steepness 0.25. function FANN_COS
Neurons     489. Current error: 0.000000. Total error:  0.0000. Epochs 98632. Bit fail   0. candidate steepness 0.75. function FANN_COS
Neurons     499. Current error: 0.000000. Total error:  0.0000. Epochs 100647. Bit fail   0. candidate steepness 0.75. function FANN_COS_SYMMETRIC
Train outputs    Current error: 0.000000. Epochs 100848
...

Failures:

  1) Fann::Network trains and evaluate single data
     Failure/Error: (result < [0.5] && result > [0.45]).should be_true

       Expected: true
            got: false

     # spec/network_spec.cr:41

Finished in 44.85 seconds
10 examples, 1 failures, 0 errors, 0 pending

Failed examples:

crystal spec spec/network_spec.cr:34 # Fann::Network trains and evaluate single data
```

**Comment by bararchy** [comment]
Hm..... it seems you also get this: "-518969491456.0000000000" as MSE ... 
@hugoabonizio What OS\Dist are you using, what version of FANN do you have ? 

I'll take another look at the specs, you would expect Neural Network to have some variance in end result from run to run, I'll make sure it's not too wide of a variance

**Comment by hugoabonizio** [comment]
Sorry @bararchy, I was running the specs before your commit b2635f8, it seems to be fixed.

But I'm running Ubuntu 16.04 and compiled FANN using the .travis.yml described script (from its master branch).

**Comment by bararchy** [comment]
@hugoabonizio No worries, now I just need to figure out why the MSE is behaving so weird, it seems to be that way at your tests too, I'll try and make a better test to check it out

**Comment by hugoabonizio** [comment]
@bararchy Thank you for your work on crystal-fann, let me know if there's something I can make to help finding the issue :smile:

**Comment by bararchy** [comment]
Yey, got it fixed via: 09fe537dda9ca8e98e1ea80653fc20508bf76d6d  
looking at fann headers it seems that even though the lib gen made MSE to be Float64, it's actually Float32.

Closing :)
